### PR TITLE
fix(list-users): return partial results when deadline is met instead of internal error

### PR DESCRIPTION
**Pull request by jon-whit**
## Description
This PR changes the ListUsers behavior such that when the ListUsers deadline is met we return the results we accumulated up to that point and avoid surfacing an internal error.

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/storage/test/storage.go [hunk 1]
```diff
@@ -53,3 +56,34 @@ func RunAllTests(t *testing.T, ds storage.OpenFGADatastore) {
 	// Stores.
 	t.Run("TestStore", func(t *testing.T) { StoreTest(t, ds) })
 }
+
+// BootstrapFGAStore is a utility which can used to quickly bootstrap an OpenFGADatastore
+// with an FGA model and relationship tuples. It returns the (store_id, model_id), respectively.
+func BootstrapFGAStore(
```
Comment (jon-whit) [inline pkg/storage/test/storage.go:62]: Bit tangential, but added this as per prior conversations and to assist with making these kinds of changes easier next time around.

File: pkg/tuple/tuple.go [hunk 2]
```diff
@@ -325,3 +325,64 @@ func IsTypedWildcard(s string) bool {
...
 }
+
+// MustParseTupleString attempts to parse a relationship tuple specified
+// in string notation and return the protobuf TupleKey for it. If parsing
+// of the string fails this  function will panic. It is meant for testing
+// purposes.
+//
+// Given string 'document:1#viewer@user:jon', return the protobuf TupleKey
+// for it.
+func MustParseTupleString(s string) *openfgav1.TupleKey {
```
Comment (jon-whit) [inline pkg/tuple/tuple.go:336]: Introduced the helpers here to help us start using the relationship tuple notation `object#relation@user` in our codebase. I wanted to do so in this PR, so I just introduced these in this.

File: pkg/server/commands/listusers/list_users_rpc.go [hunk 3]
```diff
@@ -178,22 +177,23 @@ func (l *listUsersQuery) ListUsers(
...
-	select {
-	case err := <-expandErrCh:
-		telemetry.TraceError(span, err)
-		return nil, err
-	case <-done:
-		cancelContextIfMaxResultsMet()
-		break
+	err := <-expandErrCh
+	if err != nil {
+		if !errors.Is(err, context.DeadlineExceeded) {
```
Comment (jon-whit) [inline pkg/server/commands/listusers/list_users_rpc.go:188]: One thing I want/need to add a test for here as well. If we're getting `context.DeadlineExceeded` from the storage layer, then I want to treat it differently.

I will add a test for this to make sure we're treating it properly. I think the way the code is written right now it will skip the internal error.

File: pkg/storage/test/storage.go [hunk 4]
```diff
@@ -53,3 +56,34 @@ func RunAllTests(t *testing.T, ds storage.OpenFGADatastore) {
...
+
+	tuples := tuple.MustParseTupleStrings(tupleStrs...)
+
+	batchSize := ds.MaxTuplesPerWrite()
+	for batch := 0; batch < len(tuples); batch += batchSize {
+		batchEnd := batch + batchSize
+
+		if batchEnd > len(tuples) {
+			batchEnd = len(tuples)
+		}
```
Comment (miparnisari) [inline pkg/storage/test/storage.go:82]: Nit: `batchEnd := math.Min(batch + batchSize, len(tuples)`

File: pkg/tuple/tuple_test.go [hunk 5]
```diff
@@ -468,3 +468,75 @@ func TestTypedPublicWildcard(t *testing.T) {
...
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tuple, err := ParseTupleString(test.str)
+			require.Equal(t, test.expectedErr, err != nil)
+
+			if !test.expectedErr {
+				require.NotNil(t, tuple)
+				require.Equal(t, test.tuple, tuple)
+			}
```
Comment (miparnisari) [inline pkg/tuple/tuple_test.go:539]:
```suggestion
if test.expectedErr != nil {
	require.ErrorIs(t, actualError, test.expectedErr)
} else {
	require.NotNil(t, tuple)
	require.Equal(t, test.tuple, tuple)
}
```

File: pkg/tuple/tuple_test.go [hunk 6]
```diff
@@ -468,3 +468,75 @@ func TestTypedPublicWildcard(t *testing.T) {
...
 }
+
+func TestParseTupleString(t *testing.T) {
+	tests := []struct {
+		name        string
+		str         string
+		tuple       *openfgav1.TupleKey
+		expectedErr bool
+	}{
+		{
```
Comment (miparnisari) [inline pkg/tuple/tuple_test.go:479]: can you add one test for `user:*` please?

File: pkg/tuple/tuple.go [hunk 7]
```diff
@@ -325,3 +325,64 @@ func IsTypedWildcard(s string) bool {
...
+	return tuples
+}
+
+// ParseTupleString attempts to parse a relationship tuple specified
+// in string notation and return the protobuf TupleKey for it. If parsing
+// of the string fails this  function returns an err.
+//
+// Given string 'document:1#viewer@user:jon', return the protobuf TupleKey
+// for it or an error.
+func ParseTupleString(s string) (*openfgav1.TupleKey, error) {
```
Comment (miparnisari) [inline pkg/tuple/tuple.go:360]: Thanks for adding this 🙇‍♀️  I preferred if you submitted to `main` though, so as not to make ListUsers PR even bigger 😅

File: pkg/storage/test/storage.go (hunk 1 above)
Comment (miparnisari) [inline pkg/storage/test/storage.go:62]:
```suggestion
// BootstrapFGAStore is a utility to write an FGA model and relationship tuples to a datastore. 
// It doesn't validate the model or the tuples.
// It returns the (store_id, model_id), respectively.
func BootstrapFGAStore(
```

File: pkg/server/commands/listusers/list_users_rpc.go [hunk 8]
```diff
@@ -178,22 +177,23 @@ func (l *listUsersQuery) ListUsers(
...
+			telemetry.TraceError(span, err)
+			return nil, err
+		}
 	}
+
+	<-done
```
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:194]: Why did you remove the `select` statement? That will neglect the `max_results` setting.

File: pkg/server/commands/listusers/list_users_rpc.go [hunk 9]
```diff
@@ -178,22 +177,23 @@ func (l *listUsersQuery) ListUsers(
...
 	go func() {
-		defer close(foundUsersCh)
 		internalRequest := fromListUsersRequest(req, &datastoreQueryCount)
-		if err := l.expand(cancellableCtx, internalRequest, foundUsersCh); err != nil {
-			expandErrCh <- err
-			return
-		}
+		err := l.expand(cancellableCtx, internalRequest, foundUsersCh)
+		close(foundUsersCh)
+		expandErrCh <- err
```
Comment (miparnisari) [inline pkg/server/commands/listusers/list_users_rpc.go:183]: I don't like the idea of sending `nil`s through a channel. That can cause a panic if the receiver forgets to check for nil

File: pkg/server/commands/listusers/list_users_rpc.go (hunk 3 above)
Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc.go:188]: Reminder – if we revert to a swtich (as mentioned above), we still need to nil-check the err before tracing it.

File: pkg/tuple/tuple_test.go (hunk 5 above)
Comment (jon-whit) [reply pkg/tuple/tuple_test.go:539]: We're not returning sentinel errors, so `ErrorIs` doesn't work here. The intent was not to test the error message as that is subject to change, but to assert that an error is raised. So that was the intent of the bool for `test.expectedErr` vs error type.

File: pkg/server/commands/listusers/list_users_rpc.go (hunk 8 above)
Comment (jon-whit) [reply pkg/server/commands/listusers/list_users_rpc.go]: I thought I had the control flow covered correctly, but I was wrong. I've updated that channel's variable name so it's more indicative of it's relevance and have added it back in 👍

File: pkg/server/commands/listusers/list_users_rpc.go (hunk 9 above)
Comment (jon-whit) [reply pkg/server/commands/listusers/list_users_rpc.go]: Changed this so that we only send `if err != nil`

File: pkg/server/commands/listusers/list_users_rpc.go (hunk 3 above)
Comment (jon-whit) [reply pkg/server/commands/listusers/list_users_rpc.go]: Adjust the control flow. Now we only send on the channel if `if err != nil`

File: pkg/tuple/tuple.go (hunk 7 above)
Comment (jon-whit) [reply pkg/tuple/tuple.go:360]: My thoughts there - the tests utilizing these changes are only applicable to the scope of this PR right now, but I'm happy to move it to main ahead of the curve if that's the preference. Sounds like that is, so I'll just bite the bullet and do so.

Comment (jon-whit) [reply pkg/tuple/tuple.go:360]: https://github.com/openfga/openfga/pull/1586 - moved over to a PR targeting `main`.

File: pkg/tuple/tuple_test.go (hunk 6 above)
Comment (jon-whit) [reply pkg/tuple/tuple_test.go:479]: Added this test https://github.com/openfga/openfga/pull/1586/files#diff-57a1b9934b1de403d56fdc98ba5b12686658b098e57e021f371a5dff321f0652R387-R391 in the PR targeting `main`.

File: pkg/server/commands/listusers/list_users_rpc.go (hunk 8 above)
Comment (jon-whit) [reply pkg/server/commands/listusers/list_users_rpc.go]: Fixed 👍

**Review by willvedd — APPROVED** [review]
Confirmed that this works as designed 👍
